ledger/complete/wal/wal.go

```diff
@@ -19,33 +20,61 @@ import (
 const SegmentSize = 32 * 1024 * 1024
 
 type DiskWAL struct {
+	ctx    context.Context
+	cancel context.CancelFunc
+	done   chan struct{}
+
 	wal            *prometheusWAL.WAL
 	paused         bool
 	forestCapacity int
 	pathByteSize   int
 	log            zerolog.Logger
-	// disk size reading can be time consuming, so limit how often its read
-	diskUpdateLimiter *time.Ticker
-	metrics           module.WALMetrics
-	dir               string
+	metrics        module.WALMetrics
+	dir            string
 }
 
 // TODO use real logger and metrics, but that would require passing them to Trie storage
 func NewDiskWAL(logger zerolog.Logger, reg prometheus.Registerer, metrics module.WALMetrics, dir string, forestCapacity int, pathByteSize int, segmentSize int) (*DiskWAL, error) {
-	w, err := prometheusWAL.NewSize(logger, reg, dir, segmentSize, false)
+	wal, err := prometheusWAL.NewSize(logger, reg, dir, segmentSize, false)
 	if err != nil {
 		return nil, err
 	}
-	return &DiskWAL{
-		wal:               w,
-		paused:            false,
-		forestCapacity:    forestCapacity,
-		pathByteSize:      pathByteSize,
-		log:               logger,
-		diskUpdateLimiter: time.NewTicker(5 * time.Second),
-		metrics:           metrics,
-		dir:               dir,
-	}, nil
+
+	ctx, cancel := context.WithCancel(context.TODO())
+	w := &DiskWAL{
+		ctx:    ctx,
+		cancel: cancel,
+		done:   make(chan struct{}),
+
+		wal:            wal,
+		paused:         false,
+		forestCapacity: forestCapacity,
+		pathByteSize:   pathByteSize,
+		log:            logger,
+		metrics:        metrics,
+		dir:            dir,
+	}
+	go w.emitMetrics()
+	return w, nil
+}
+
+func (w *DiskWAL) emitMetrics() {
+	defer close(w.done)
+
+	t := time.NewTicker(10 * time.Second)
```

Comment:
Actually, I probably should move it to 15 seconds (our prometheus resolution) or even higher since it is quite expensive, but produces a lot of system calls (it needs to do one `stat(2)` per file in directory) and we have a separate metric for the partition-wide disk usage that we alert on.